Inject `hop_to_executor` after `self` is fully initialized in unrestricted actor inits

kavon opened the pull request:
Since `async` actor initializers (designated ones) are allowed to make unrestricted use of `self` after the point where it is fully-initialized, we need to hop to the `self` executor after those points in order to prevent races in the init. 

The design of initializer declarations in Swift means that there are multiple such fully-initialized points. To find them, you need to use a flow-sensitive analysis of stores to properties, etc. 

We already do this analysis in DefiniteInitialization for classes, etc. So, this PR takes advantage of the information DI already computes to find these points and insert the hop instructions, thus plugging the data-race.

Resolves rdar://78790683

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+
+#ifndef NDEBUG
+    // dump the data-structure we just built
+    for (auto &entry : initializingBlocks) {
```
Please hide this behind LLVM_DEBUG. People compile with asserts all the time and they shouldn't have to see this.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+  // Tracks status of each element of TheMemory as we scan through the block.
+  AvailabilitySet localAvail;
+
+  for (auto const& entry : initializingBlocks) {
```
Can you run this through git-clang-format?

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+    // set-up initial availability at the block's entry-point.
+    localAvail = entry.getSecond();
+
+    auto BBI = block->begin(); // our cursor and eventual insertion-point.
```
Please us `bbi` and `bbe`.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+
+    // TODO: obtain `self` from the entry-block and insert the hop.
+    // probably need to emit a borrow and then clean up injectHopAfterBuiltinInit, etc.
+    llvm::dbgs() << "would insert hop_to_executor before: ";
```
All of these dbgs things need to be guarded with LLVM_DEBUG.

kavon replied:
OK. that line will be gone soon though.

gottesmm reviewed (commented):
Some quick structural comments.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+
+  Map initializingBlocks;
+  {
+    llvm::SmallDenseSet<SILBasicBlock *> seenBlocks; // processed blocks
```
Please do not use your own SmallDenseSet. Instead use this: https://github.com/apple/swift/blob/7123d2614b5f222d03b3762cb110d27a9dd98e24/include/swift/SIL/BasicBlockBits.h#L160.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+    return injectHopAfterBuiltinInit(ctor,F.getEntryBlock());
+
+  // A mapping from initializing blocks to their in-availability set.
+  using Map = llvm::SmallDenseMap<SILBasicBlock*, AvailabilitySet, 8>;
```
If you are going to do this, please extract this onto the lifetime checker and clear it every time. That way if we go big, we never need to go big again and can just reuse the memory.

kavon replied:
Went with Erik's suggestion to make this a locally-allocated vector for now.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+  if (isRestrictedActorInitSelf())
+    return;
+
+  // Even if there are no stored properties to initialize, we still need a hop.
```
Can you explain here /why/ you are inserting the hop after the builtin inst. That is the interesting part.

gottesmm reviewed (commented):
Some more things.

eeckstein commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+  // Step 1: Find initializing blocks, which are blocks that contain a store
+  // to TheMemory that fully-initializes it, and build the Map.
+
+  Map initializingBlocks;
```
This doesn't need to be a map. It can be an array of (SILBasicBlock,  AvailabilitySet) tuples.
This also solves the problem of iterating over this data structure: iterating over a  pointer hash map is very dangerous because it can be easily introducing non-deterministic behavior.

eeckstein commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+  // Tracks status of each element of TheMemory as we scan through the block.
+  AvailabilitySet localAvail;
+
+  for (auto const& entry : initializingBlocks) {
```
Iterating over a  pointer hash map is very dangerous because it can be easily introducing non-deterministic behavior. E.g, in this case, the order of the use list of the `hop_to_executor` operand would be non-deterministic.

Even if you make `initializingBlocks` an array, it would not solve the problem, because you are computing `initializingBlocks` by iterating over the `NonLoadUses` hash map.

You could just iterate over all basic blocks in the function to get a deterministic order. Complexity-wise this would be okay, because `injectActorHops()` is only done a single time for an actor constructor.

kavon replied:
Gotcha. I'll go over all the blocks and ask if they have non-load uses, and then scan each one, so that its more deterministic.

kavon committed:
inject hop_to_executor in async actor inits

Because `self` can be used unrestricted after it is
fully-initialized in an async actor init, we perform
a hop_to_executor(self) immediately after `self`
is fully-initialized on all paths within the
initializer.

If we did not, then code could race with the
initializer if it, e.g., spawns a task from
the initializer and mutates actor state that
the initializer then reads. By hopping to the
executor, we prevent that task from running
until _after_ the initializer completes.

eeckstein commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+
+      // Have we found the instruction that initializes TheMemory?
+      if (localAvail.isAllYes()) {
+        bbi++; // bump to next instruction for an insertion point.
```
Is it always guaranteed that `inst` is _not_ a terminator instruction? E.g. can `use` be an inout argument of a `try_apply`?
Because if it's a terminator instruction, you cannot insert the `hop_to_executor` after it.
Instead you can use `SILBuilderWithScope::insertAfter`.

kavon replied:
I decided to use `insertAfter` but also added an assert ensuring its not a terminator, because `insertAfter` technically doesn't support some kinds of successors (i.e., those with multiple predecessors).

It should be always guaranteed because I don't see any kinds of terminators that do a store, and `inout` uses are banned before the stored property is initialized.

kavon replied:
Good point though thank you for catching this.

eeckstein commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+    }
+
+    // Make sure we found the initializing use of TheMemory.
+    if (!localAvail.isAllYes())
```
Is it intention to not just use an `assert` here?

kavon replied:
It might actually not be a failure, because I need to see what happens if the block has no successors and doesn't fully initialize `self`, i.e., it's a failable init and returns `nil`.

kavon replied:
I made it an assert since that doesn't actually happen.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+static void injectHopToExecutorBefore(SILLocation loc,
+                                      SILBasicBlock::iterator &insertPt,
+                                      SILValue actor, bool needsBorrow) {
+  SILBuilder B(insertPt);
```
This needs to be a SILBuilderWithScope. We no longer use SILBuilder. We just have never finished the transition. Also please use b or builder.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+  SILBuilder B(insertPt);
+
+  LLVM_DEBUG(llvm::dbgs() << "hop-injector: inserting hop_to_executor before ");
+  LLVM_DEBUG(insertPt->dump());
```
Instead of using dump here, please just pass *insertPt to the line before. dump is meant to be just used in the debugger.

gottesmm replied:
Example: https://github.com/apple/swift/blob/191dc14c59a96b2c9a38540e3f3e8624e1e7404d/lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp#L3177

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+
+  if (needsBorrow) {
+    auto *borrow = B.createBeginBorrow(loc, actor);
+    assert(borrow->getNumResults() == 1);
```
Borrow will always have one result. Please remove this.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+  }
+
+  // A collection of pairs: an initializing block and its in-availability set.
+  using Vec = llvm::SmallVector<std::pair<SILBasicBlock *, AvailabilitySet>, 8>;
```
Is this needed? Do we get more than 80 columns here? It really doesn't add anything. Also can you extract this into a class level data structure that we can clear. That way if we go big, we can just reuse memory.

kavon replied:
I just like type aliases for long types, and hadn't settled on a good name for it yet.

Do you really want me to move it into the LifetimeChecker for optimization purposes? What if I just made it a `static` decl in the function? Would that avoid the allocation only until the first time `injectHops` is called?

kavon replied:
I have instead simply eliminated the data structure 😄

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -1786,8 +1941,7 @@ bool LifetimeChecker::diagnoseReturnWithoutInitializingStoredProperties(
 
-bool LifetimeChecker::isRestrictedActorInitSelf(const DIMemoryUse& Use,
-                                            ActorInitKind *Kind) const {
+bool LifetimeChecker::isRestrictedActorInitSelf(ActorInitKind *Kind) const {
```
I think this snuck in, but can you please use camelCase. This is just an old file that needs to be updated a little. New code should use camelCase.

gottesmm commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -783,6 +786,155 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+}
+
+void LifetimeChecker::injectActorHops() {
+  auto ctor = TheMemory.isActorInitSelf();
```
Please do not return a pointer from an is method. It is misleading since is methods are assumed to return something bool by convention. Keeping in mind that I do not think that you actually need to use this here, for future methods please use a get method, like TheMemory.getActorInitSelf(). If this an is that you really want to use, please return a bool. The reason why I don't think you need the pointer here is the correct thing to use in the inject method (where you are passing this) is just SILBuilderWithScope at the insert pt.

kavon replied:
That method is used in other places where the pointer is needed. I'll just rename it to use `get` in front instead.

gottesmm reviewed (commented):
Some initial comments. Need to go through again for the algorithm.

kavon committed:
SIL test for hop_to_executor in async actor inits

kavon commented:
@swift-ci please smoke test

kavon committed:
runtime regression test for async actor init

kavon commented:
There's a complication right now: a `hop_to_executor` cannot appear within a `begin_access`/`end_access` pair. AFAIK its not trivial to determine where these accesses are. Might need to mark the definite init locations and then break out the injection into a follow-up pass.

For now, trying something hacky to see what breaks: blindly inject at the beginning of every async init.

kavon commented:
@swift-ci please smoke test

kavon committed:
add assertion to catch AccessSet usage errors

I ran into this while implementing
async actor inits; where I was emitting
an actor hop in the middle of
an access region, which caused the
access set to be unexpectedly empty.

kavon committed:
insert hops after end_access when possible

atrick commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -810,15 +812,57 @@ static void injectHopToExecutorAfter(SILLocation loc,
+  auto getBeginAccess = [](SILValue v) -> BeginAccessInst * {
+    // skip static access ranges, since they can't trigger any issues.
+    if (auto *access = dyn_cast<BeginAccessInst>(getAccessScope(v)))
+      if (access->getEnforcement() != SILAccessEnforcement::Static)
```
It doesn't hurt to check for non-Static enforcement, but only creates a special case. I think we're better off handling all access scopes uniformly and maintaining the invariant that the hop does not occur within the access.

kavon commented:
@swift-ci please test

atrick commented on stdlib/public/runtime/Exclusivity.cpp:
```diff
@@ -202,6 +202,7 @@ class AccessSet {
   }
 
   void remove(Access *access) {
```
How does this new assert help? If the access set is empty, you'll always hit the assert below.

kavon replied:
If it's empty, you'll trigger a seg-fault on the `cur->getNext()` (the main caller doesn't call this method if `access` is `nullptr`). When I was debugging this with a Release+Asserts compiler, it doesn't directly point to that line, since it gets inlined / optimized somewhat. That's why I added it. 

Do you think I should remove this assert?

atrick replied:
I see. I think the assert is good, just didn't understand that it was avoiding a non-assert crash.

atrick commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -810,15 +812,54 @@ static void injectHopToExecutorAfter(SILLocation loc,
+  };
+
+  //////
+  // FIXME: It is invalid for a hop to appear within an access scope, since the
```
I'm not sure what this FIXME is for. Should we say:
This is a best-effort to avoid placing hops within an access, because we would prefer the access set to be empty.
But the hop may still be emitted in an overlapping access in theory, and the runtime can handle that?

It would be awesome to come up with a test case for overlapping access, but I don't know how to do it given that properties can neither be initialized using 'inout' nor with a closure

kavon replied:
Okay, I'll change it to a NOTE explaining the rationale for doing this extra work.

atrick reviewed (commented):
This looks good to me (as a way to hide the problem for now). Is there value in adding the minimal reproducer as a runtime test?

kavon commented:
@swift-ci please smoke test

kavon committed:
more coverage for SILGen async_initializer test

kavon commented:
@swift-ci please smoke test and merge

kavon commented:
@swift-ci please smoke test and merge
